tests/Unit/Evolution/Systems/ScalarWave/Test_Characteristics.cpp

```diff
@@ -0,0 +1,374 @@
+  auto unit_normal_one_form =
+      make_with_value<tnsr::i<DataVector, Dim, Frame::Inertial>>(
+          x, 1. / sqrt(Dim));
+
+  // Fundamental fields (psi, pi, phi) have already been computed locally.
+  // Now, check that these locally computed fields match returned ones
+  // First, compute characteristic fields locally
+  const auto phi_dot_normal = dot_product(unit_normal_one_form, phi_expected);
+
+  tnsr::i<DataVector, Dim, Frame::Inertial> phi_dot_projection_tensor{
+      DataVector(n_pts)};
+  for (size_t i = 0; i < Dim; ++i) {
+    phi_dot_projection_tensor.get(i) =
+        phi_expected.get(i) - unit_normal_one_form.get(i) * get(phi_dot_normal);
+  }
+
+  const auto& vpsi = psi_expected;
+  const auto& vzero = phi_dot_projection_tensor;
+  const Scalar<DataVector> vplus{get(pi_expected) + get(phi_dot_normal) -
+                                 get(gamma_2) * get(psi_expected)};
+  const Scalar<DataVector> vminus{get(pi_expected) - get(phi_dot_normal) -
+                                  get(gamma_2) * get(psi_expected)};
+  // Second, obtain fundamental fields using compute tag
+  {
+    const auto fields =
+        ScalarWave::EvolvedFieldsFromCharacteristicFieldsCompute<Dim>::function(
+            gamma_2, vpsi, vzero, vplus, vminus, unit_normal_one_form);
+    const auto& psi = get<ScalarWave::Psi>(fields);
+    const auto& pi = get<ScalarWave::Pi>(fields);
+    const auto& phi = get<ScalarWave::Phi<Dim>>(fields);
+
+    CHECK_ITERABLE_APPROX(psi_expected, psi);
+    CHECK_ITERABLE_APPROX(pi_expected, pi);
+    CHECK_ITERABLE_APPROX(phi_expected, phi);
+  }
+  // Third, obtain fundamental fields using function
+  {
+    const auto fields = ScalarWave::evolved_fields_from_characteristic_fields(
+        gamma_2, vpsi, vzero, vplus, vminus, unit_normal_one_form);
+    const auto& psi = get<ScalarWave::Psi>(fields);
+    const auto& pi = get<ScalarWave::Pi>(fields);
+    const auto& phi = get<ScalarWave::Phi<Dim>>(fields);
+
+    CHECK_ITERABLE_APPROX(psi_expected, psi);
+    CHECK_ITERABLE_APPROX(pi_expected, pi);
+    CHECK_ITERABLE_APPROX(phi_expected, phi);
+  }
+}
+}  // namespace
+
+SPECTRE_TEST_CASE("Unit.Evolution.Systems.ScalarWave.Characteristics",
+                  "[Unit][Evolution]") {
+  pypp::SetupLocalPythonEnvironment local_python_env{
+      "Evolution/Systems/ScalarWave/"};
+
+  test_characteristic_speeds<1>();
+  test_characteristic_speeds<2>();
+  test_characteristic_speeds<3>();
+
+  test_characteristic_fields<1>();
+  test_characteristic_fields<2>();
+  test_characteristic_fields<3>();
+
+  test_evolved_from_characteristic_fields<1>();
+  test_evolved_from_characteristic_fields<2>();
+  test_evolved_from_characteristic_fields<3>();
+
+  // Test characteristics against 3D spherical wave
+  const size_t grid_size = 8;
+  const std::array<double, 3> lower_bound{{0.82, 1.22, 1.32}};
+  const std::array<double, 3> upper_bound{{0.78, 1.18, 1.28}};
+
+  const ScalarWave::Solutions::RegularSphericalWave solution(
```

Comment:
Non-binding: I suggest renaming this to `spherical_wave_solution`

Reply:
changed